substrate/primitives/runtime/src/proving_trie/base2.rs

```diff
@@ -0,0 +1,246 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types for a base-2 merkle tree used for checking and generating proofs within the
+//! runtime. The `binary-merkle-tree` crate exposes all of these same functionality (and more), but
+//! this library is designed to work more easily with runtime native types, which simply need to
+//! implement `Encode`/`Decode`.
+
+use super::{ProvingTrie, TrieError};
+use crate::{Decode, DispatchError, Encode};
+use binary_merkle_tree::{merkle_proof, merkle_root, MerkleProof};
+use codec::MaxEncodedLen;
+use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+/// A helper structure for building a basic base-2 merkle trie and creating compact proofs for that
+/// trie.
+pub struct BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+{
+	// Deduplicated and flattened list of key value pairs.
+	db: BTreeMap<Key, Value>,
+	root: Hashing::Out,
+	_phantom: core::marker::PhantomData<(Key, Value)>,
+}
+
+impl<Hashing, Key, Value> ProvingTrie<Hashing, Key, Value> for BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Hashing::Out: Encode + Decode + MaxEncodedLen,
+	Key: Encode + Decode + Ord,
+	Value: Encode + Decode + Clone,
+{
+	/// Create a new instance of a `ProvingTrie` using an iterator of key/value pairs.
+	fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>,
+	{
+		let mut db = BTreeMap::default();
+		for (key, value) in items.into_iter() {
+			db.insert(key, value);
+		}
+		let root = merkle_root::<Hashing, _>(db.iter().map(|item| item.encode()));
+		Ok(Self { db, root, _phantom: Default::default() })
+	}
+
+	/// Access the underlying trie root.
+	fn root(&self) -> &Hashing::Out {
+		&self.root
+	}
+
+	/// Query a value contained within the current trie. Returns `None` if the
+	/// nodes within the current `db` are insufficient to query the item.
+	fn query(&self, key: &Key) -> Option<Value> {
+		self.db.get(&key).cloned()
+	}
+
+	/// Create a compact merkle proof needed to prove a single key and its value are in the trie.
+	/// Returns `None` if the nodes within the current `db` are insufficient to create a
+	/// proof.
+	fn create_proof(&self, key: &Key) -> Result<Vec<u8>, DispatchError> {
+		let mut encoded = Vec::with_capacity(self.db.len());
+		let mut found_index = None;
+
+		// Find the index of our key, and encode the (key, value) pair.
+		for (i, (k, v)) in self.db.iter().enumerate() {
+			// If we found the key we are looking for, save it.
+			if k == key {
+				found_index = Some(i);
+			}
+
+			encoded.push((k, v).encode());
+		}
+
+		let index = found_index.ok_or(TrieError::IncompleteDatabase)?;
+		let proof = merkle_proof::<Hashing, Vec<Vec<u8>>, Vec<u8>>(encoded, index as u32);
```

Comment:
Could probably be optimized later to not have quadratic complexity, but i guess its fine for the first version.

Reply:
This should be linear? If we can optimize it, let's do it, this shouldn't be crazy expensive function.

Reply:
Ah yea 🤦‍♂️ `merkle_proof` is looping again, but only over all elements, not for each one.